// @stylexjs/shared/lib/preprocess-rules/PreRule.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreRuleSet = exports.PreRule = exports.PreIncludedStylesRule = exports.NullPreRule = void 0;
var _convertToClassName = require("../convert-to-className");
var _objectUtils = require("../utils/object-utils");
class NullPreRule {
  compiled(_options) {
    return [null];
  }
  equals(other) {
    return other instanceof NullPreRule;
  }
}
exports.NullPreRule = NullPreRule;
class PreIncludedStylesRule {
  constructor(IncludedStyles) {
    this.includedStyles = IncludedStyles;
  }
  equals(other) {
    return other instanceof PreIncludedStylesRule && this.includedStyles === other.includedStyles;
  }
  compiled(_options) {
    return this.includedStyles;
  }
}
exports.PreIncludedStylesRule = PreIncludedStylesRule;
const stringComparator = (a, b) => {
  if (a === 'default') {
    return -1;
  }
  if (b === 'default') {
    return 1;
  }
  return a.localeCompare(b);
};
class PreRule {
  constructor(property, value, pseudos, atRules) {
    this.property = property;
    this.value = value;
    this.pseudos = pseudos ? (0, _objectUtils.arraySort)(pseudos, stringComparator) : [];
    this.atRules = atRules ? (0, _objectUtils.arraySort)(atRules) : [];
  }
  compiled(options) {
    const [_key, className, rule] = (0, _convertToClassName.convertStyleToClassName)([this.property, this.value], this.pseudos ?? [], this.atRules ?? [], options);
    return [[className, rule]];
  }
  equals(other) {
    if (!(other instanceof PreRule)) {
      return false;
    }
    const valuesEqual = Array.isArray(this.value) && Array.isArray(other.value) ? (0, _objectUtils.arrayEquals)(this.value, other.value) : this.value === other.value;
    return this.property === other.property && valuesEqual && (0, _objectUtils.arrayEquals)(this.pseudos, other.pseudos) && (0, _objectUtils.arrayEquals)(this.atRules, other.atRules);
  }
}
exports.PreRule = PreRule;
class PreRuleSet {
  constructor(rules) {
    this.rules = rules;
  }
  static create(rules) {
    const flatRules = rules.flatMap(rule => rule instanceof PreRuleSet ? rule.rules : [rule]);
    if (flatRules.length === 0) {
      return new NullPreRule();
    }
    if (flatRules.length === 1) {
      return flatRules[0];
    }
    return new PreRuleSet(flatRules);
  }
  compiled(options) {
    const styleTuple = this.rules.flatMap(rule => rule.compiled(options)).filter(Boolean);
    return styleTuple.length > 0 ? styleTuple : [null];
  }
  equals(other) {
    if (!(other instanceof PreRuleSet)) {
      return false;
    }
    if (this.rules.length !== other.rules.length) {
      return false;
    }
    return (0, _objectUtils.arrayEquals)(this.rules, other.rules, (a, b) => a.equals(b));
  }
}
exports.PreRuleSet = PreRuleSet;
