// @tanstack/router-plugin/dist/cjs/code-splitter.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_path = require("node:path");
const node_url = require("node:url");
const compilers = require("./compilers.cjs");
const config = require("./config.cjs");
const constants = require("./constants.cjs");
function capitalizeFirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}
function fileIsInRoutesDirectory(filePath, routesDirectory) {
  const routesDirectoryPath = node_path.isAbsolute(routesDirectory) ? routesDirectory : node_path.join(process.cwd(), routesDirectory);
  return filePath.startsWith(routesDirectoryPath);
}
const bannedBeforeExternalPlugins = [
  {
    identifier: "@react-refresh",
    pkg: "@vitejs/plugin-react",
    usage: "viteReact()",
    frameworks: ["vite"]
  }
];
class FoundPluginInBeforeCode extends Error {
  constructor(externalPlugin, framework) {
    super(`We detected that the '${externalPlugin.pkg}' was passed before '@tanstack/router-plugin'. Please make sure that '@tanstack/router-plugin' is passed before '${externalPlugin.pkg}' and try again: 
e.g.
plugins: [
  TanStackRouter${capitalizeFirst(framework)}(), // Place this before ${externalPlugin.usage}
  ${externalPlugin.usage},
]
`);
  }
}
const PLUGIN_NAME = "unplugin:router-code-splitter";
const JoinedSplitPrefix = constants.splitPrefix + ":";
const unpluginRouterCodeSplitterFactory = (options = {}, { framework }) => {
  const debug = Boolean(process.env.TSR_VITE_DEBUG);
  let ROOT = process.cwd();
  let userConfig = options;
  const handleSplittingFile = (code, id) => {
    if (debug) console.info("Splitting route: ", id);
    const compiledVirtualRoute = compilers.compileCodeSplitVirtualRoute({
      code,
      root: ROOT,
      filename: id
    });
    if (debug) console.info("");
    if (debug) console.info("Split Output");
    if (debug) console.info("");
    if (debug) console.info(compiledVirtualRoute.code);
    if (debug) console.info("");
    if (debug) console.info("");
    if (debug) console.info("");
    return compiledVirtualRoute;
  };
  const handleCompilingFile = (code, id) => {
    if (debug) console.info("Handling createRoute: ", id);
    const compiledReferenceRoute = compilers.compileCodeSplitReferenceRoute({
      code,
      root: ROOT,
      filename: id
    });
    if (debug) console.info("");
    if (debug) console.info("Compiled Output");
    if (debug) console.info("");
    if (debug) console.info(compiledReferenceRoute.code);
    if (debug) console.info("");
    if (debug) console.info("");
    if (debug) console.info("");
    return compiledReferenceRoute;
  };
  return {
    name: "router-code-splitter-plugin",
    enforce: "pre",
    resolveId(source) {
      var _a;
      if (!((_a = userConfig.experimental) == null ? void 0 : _a.enableCodeSplitting)) {
        return null;
      }
      if (source.startsWith(constants.splitPrefix + ":")) {
        return source.replace(constants.splitPrefix + ":", "");
      }
      return null;
    },
    async transform(code, id) {
      var _a;
      if (!((_a = userConfig.experimental) == null ? void 0 : _a.enableCodeSplitting)) {
        return null;
      }
      const url = node_url.pathToFileURL(id);
      url.searchParams.delete("v");
      id = node_url.fileURLToPath(url).replace(/\\/g, "/");
      if (id.includes(constants.splitPrefix)) {
        return await handleSplittingFile(code, id);
      } else if (fileIsInRoutesDirectory(id, userConfig.routesDirectory) && (code.includes("createRoute(") || code.includes("createFileRoute("))) {
        for (const externalPlugin of bannedBeforeExternalPlugins) {
          if (!externalPlugin.frameworks.includes(framework)) {
            continue;
          }
          if (code.includes(externalPlugin.identifier)) {
            throw new FoundPluginInBeforeCode(externalPlugin, framework);
          }
        }
        return await handleCompilingFile(code, id);
      }
      return null;
    },
    transformInclude(transformId) {
      var _a;
      if (!((_a = userConfig.experimental) == null ? void 0 : _a.enableCodeSplitting)) {
        return void 0;
      }
      let id = transformId;
      if (id.startsWith(JoinedSplitPrefix)) {
        id = id.replace(JoinedSplitPrefix, "");
      }
      if (fileIsInRoutesDirectory(id, userConfig.routesDirectory) || id.includes(constants.splitPrefix)) {
        return true;
      }
      return false;
    },
    vite: {
      configResolved(config$1) {
        ROOT = config$1.root;
        userConfig = config.getConfig(options, ROOT);
      }
    },
    rspack(compiler) {
      ROOT = process.cwd();
      compiler.hooks.beforeCompile.tap(PLUGIN_NAME, (self) => {
        self.normalModuleFactory.hooks.beforeResolve.tap(
          PLUGIN_NAME,
          (resolveData) => {
            if (resolveData.request.includes(JoinedSplitPrefix)) {
              resolveData.request = resolveData.request.replace(
                JoinedSplitPrefix,
                ""
              );
            }
          }
        );
      });
      userConfig = config.getConfig(options, ROOT);
    },
    webpack(compiler) {
      var _a;
      ROOT = process.cwd();
      compiler.hooks.beforeCompile.tap(PLUGIN_NAME, (self) => {
        self.normalModuleFactory.hooks.beforeResolve.tap(
          PLUGIN_NAME,
          (resolveData) => {
            if (resolveData.request.includes(JoinedSplitPrefix)) {
              resolveData.request = resolveData.request.replace(
                JoinedSplitPrefix,
                ""
              );
            }
          }
        );
      });
      userConfig = config.getConfig(options, ROOT);
      if (((_a = userConfig.experimental) == null ? void 0 : _a.enableCodeSplitting) && compiler.options.mode === "production") {
        compiler.hooks.done.tap(PLUGIN_NAME, (stats) => {
          console.info("✅ " + PLUGIN_NAME + ": code-splitting done!");
          setTimeout(() => {
            process.exit(0);
          });
        });
      }
    }
  };
};
exports.unpluginRouterCodeSplitterFactory = unpluginRouterCodeSplitterFactory;
